integration/common.sh

```diff
@@ -20,26 +20,51 @@ function runTest() {
 }
 
 # Wait for a server to become ready
-function waitForServerStartup() {
+waitForServerStartup() {
   # The server will 404 the request as there's no handler for it. This error doesn't matter
   # as the test will fail if the server is really not up.
-  PORT=$1
-  wget -q --spider --retry-connrefused --waitretry=1 -t ${STARTUP_WAIT_SECONDS} localhost:${PORT}
+  local port=$1
+  wget -q --spider --retry-connrefused --waitretry=1 -t ${STARTUP_WAIT_SECONDS} localhost:${port}
   # Wait a bit more to give it a chance to become actually available e.g. if Travis is slow
   sleep 2
 }
 
+killPid() {
+  local pid=$1
+  set +e
+  local count=0
+  while kill -INT ${pid} > /dev/null 2>&1; do
+    sleep 1
+    ((count++))
+    if ! ps -p ${pid} > /dev/null ; then
+      break
+    fi
+    if [ $count -gt 5 ]; then
+      echo "Now do kill -KILL ${pid}"
+      kill -KILL ${pid}
+      break
+    fi
+    echo "Retry kill -INT ${pid}"
+  done
+  set -e
+}
 
 # Clean up anything in ${TO_KILL} and ${TO_DELETE}.
-function onExit() {
-  if [[ "${TO_KILL}" ]]; then
-    echo "Killing ${TO_KILL} on exit"
-    kill -INT ${TO_KILL}
-  fi
-  if [[ "${TO_DELETE}" ]]; then
-    echo "Deleting ${TO_DELETE} on exit"
-    rm ${TO_DELETE}
-  fi
+declare -a TO_KILL
+declare -a TO_DELETE
+onExit() {
+  local pid=0
+  for pid in "${TO_KILL[@]}"
```

Comment:
FYI, you're still missing a few "for ...; do" occurrences (not counting existing code).